docs/feature/dualstack/Cilium-dualstack.md

```diff
@@ -0,0 +1,67 @@
+[`AssignAnyAvailableIPConfig`](https://github.com/Azure/azure-container-networking/blob/master/cns/restserver/ipam.go#:~:text=func%20(service%20*HTTPRestService)%20AssignAnyAvailableIPConfig(podInfo%20cns.PodInfo)%20(cns.PodIpInfo%2C%20error)%20%7B)
+## New Implementation
+
+### NNC change
+
+For the CNS to know that we are in dual stack we will use a flag that will be passed in using the NNC. This will be filled in from the DNC-RC side and then sent to the CNS over the CRD in a flag called `IPMode`. The IPMode flag will be of type IPMode and will have two options "SingleStack" which when passed in will have the CNS work as it currently does and "Dual Stack" which tell the CNS to look for two IPs instead of one.
+
+```diff
+// NodeNetworkConfigStatus defines the observed state of NetworkConfig
+type NodeNetworkConfigStatus struct {
+	AssignedIPCount   int                `json:"assignedIPCount,omitempty"`
+	Scaler            Scaler             `json:"scaler,omitempty"`
+	Status            Status             `json:"status,omitempty"`
++   IPMode            IPMode             `json:"IPmode,omitempty"`
+	NetworkContainers []NetworkContainer `json:"networkContainers,omitempty"`
+}
+```
+
+```diff
++type IPMode string
++
++const (
++	SingleStack Status = "SingleStack"
++	DualStack   Status = "DualStack"
++)
+```
+
+In the CNS will flag then be used in the [`AssignAnyAvailableIPConfig`](https://github.com/Azure/azure-container-networking/blob/master/cns/restserver/ipam.go#:~:text=func%20(service%20*HTTPRestService)%20AssignAnyAvailableIPConfig(podInfo%20cns.PodInfo)%20(cns.PodIpInfo%2C%20error)%20%7B) function to determine whether or not are creating a pod in single stack or dual stack. 
+
+### Contract IP change
+
+We will continue to send only one `IPConfigRequest` when we are running in dual stack but when in dual stack we will expect the repsonse to include a slice of `PodIpInfo`. When running the [`AssignAnyAvailableIPConfig`](https://github.com/Azure/azure-container-networking/blob/master/cns/restserver/ipam.go#:~:text=func%20(service%20*HTTPRestService)%20AssignAnyAvailableIPConfig(podInfo%20cns.PodInfo)%20(cns.PodIpInfo%2C%20error)%20%7B) function with the new flag for dual stack if we are running in dual stack we will have the function pick the first availabe IPv6 and first available IPv4 IP that it can find. These IPs will now be returned as a slice from `AssignAnyAvailableIPConfig` and used to populate the `IPConfigResponse` struct and returned back to the azure-ipam. The azure-ipam will then cycle the length of the slice to populate the message the is sent to the Cilium-CNI.
+
+```
+type IPConfigRequest struct {
+    DesiredIPAddress    string
+    PodInterfaceID      string
+    InfraContainerID    string
+    OrchestratorContext json.RawMessage
+    Ifname              string // Used by delegated IPAM
+}
+```
+
+```diff
+type IPConfigResponse struct {
+-   PodIpInfo PodIpInfo 
++   PodIpInfo []PodIpInfo 
+    Response  Response
+}
```

Comment:
actually new type is probably not required, the new api can just return the slice as top-level lists are still valid json

Reply:
When you say that I wouldn't need a new type are you suggesting that I just return a slice of the IPConfigResponse instead of the PodIPInfo within it?

Reply:
@thatmattlong IMO it would be better to create a new type so that there will be a single 'Response' indicating if the client should consume the response. Can't do that if we return slice of IPConfigResponse